shapeshift/web
# feat: ens address lookups

⚠️ Do not merge, this requires typings update currently on CR in https://github.com/shapeshift/lib/pull/315 - will bump the version in package.json when lib update is merged and published ⚠️

## Description

This adds client-side support for ENS resolution.
- Lookup (address -> domain)
- Reverse lookup (domain -> address)

Refer to closing issue for full spec.

### Notes

I added these as a concern after initial commit. It probably isn't much of a concern anymore given the memoized ENS helpers for reverse/lookups.
~~~
This is a very component-(co)located implementation of ENS reverse/lookups.
It works according to the spec but I think the design could be improved with a more redux way:
- What if we need to fetch ENS in more places in the future, are we going to call the ENS instance every time?
- What about memoization? Using an async method in place makes it impossible. 

Ideally, I think that resolutions would make more sense to be moved to their respective redux slices in the future.
That can be done in two ways, either everywhere you have an ETH address in state, you'd also have the matching ENS domain, or `null`. That'd probably require changes like `@shapeshift-lib`'s [`subscribeTxs`](https://github.com/shapeshift/lib/blob/ca9e1b5ab18996f222dabd487835a9ab4b7b818a/packages/chain-adapters/src/ethereum/EthereumChainAdapter.ts#L298) sending  the ens domain along with the address. Also, it would possibly make the affected states less blockchain-agnostic.
Or, it can be its own rtk-query slice. That would be more straightforward since you'd just need one mapping of address to domain. 

Either way, memoization would be possible with `reselect` or built into rtk-query.

**Open to recommendations on how to refactor it now, or if you guys think the implementation makes sense for now, later!**
~~~
## Notice

Before submitting a pull request, please make sure you have answered the following:


## Pull Request Type


## Issue

Closes #674

## Testing

Please outline all testing steps

1. Install latest version of`@shapeshiftoss/chain-adapters` (or checkout https://github.com/shapeshift/lib/pull/296/files, build it, and link it if new version is not published yet)
2. Try to pass a valid, resolvable ENS address as input in "Send" and complete flow to send confirm click
3. Metamask modal should appear with the ENS as destination
4. Try to pass invalid ENS addresses e.g addresses not fitting the UTS46 spec or addresses with no owner/resolver should error. The flow I've decided to go with is to have "Next" button in loading state while checking for resolution, then erroring with some error message copy
5. Have some transactions for your current address with sender that has ENS domain resolving to it, it should be displayed instead of address
6. Click receive with a connected account that has ENS domain resolving to it, a rounded tag should be displayed simillar to acceptation criteria 

## Screenshots

![image](https://user-images.githubusercontent.com/17035424/148746188-58a79100-44f0-40f6-89fa-84a1281c2ac7.png)
![image](https://user-images.githubusercontent.com/17035424/148746222-a6b1b150-2ce6-4fb0-ae7a-46ff4ca461ec.png)
![image](https://user-images.githubusercontent.com/17035424/148746264-90510a0b-c00e-4708-b45f-41163877a4b7.png)
![image](https://user-images.githubusercontent.com/17035424/148746333-b8fe35c0-def0-4fca-8161-44d5364a5f09.png)
![image](https://user-images.githubusercontent.com/17035424/148857880-695aec09-74f6-4269-9509-647525c7f88a.png)
![image](https://user-images.githubusercontent.com/17035424/149047445-c3f3be24-0c37-4611-855e-d07427f13ae8.png)
